hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheProxy.java

```diff
@@ -216,9 +219,49 @@ public V getAndPut(K key, V value, ExpiryPolicy expiryPolicy) {
     public void putAll(Map<? extends K, ? extends V> map, ExpiryPolicy expiryPolicy) {
         ensureOpen();
         validateNotNull(map);
-        // TODO implement batch putAll
-        for (Map.Entry<? extends K, ? extends V> entry : map.entrySet()) {
-            put(entry.getKey(), entry.getValue(), expiryPolicy);
+
+        int partitionCount = partitionService.getPartitionCount();
+
+        try {
+            List<Future> futures = new ArrayList<Future>(partitionCount);
+            List<Map.Entry<Data, Data>>[] entriesPerPartition = new List[partitionCount];
+
+            // First we fill entrySetPerPartition
+            for (Map.Entry<? extends K, ? extends V> entry : map.entrySet()) {
+                K key = entry.getKey();
+                V value = entry.getValue();
+                validateNotNull(key, value);
+
+                Data keyData = serializationService.toData(entry.getKey());
+                Data valueData = serializationService.toData(entry.getValue());
+
+                int partitionId = partitionService.getPartitionId(keyData);
+                List<Map.Entry<Data, Data>> entries = entriesPerPartition[partitionId];
+                if (entries == null) {
+                    entries = new ArrayList<Map.Entry<Data, Data>>();
+                    entriesPerPartition[partitionId] = entries;
+                }
+
+                entries.add(new AbstractMap.SimpleImmutableEntry<Data, Data>(keyData, valueData));
+            }
+
```

Comment:
Maybe it would be good for readibility to extract a method called 
`groupDataToPartitions(Map<? extends K, ? extends V> map)` that would return grouped data.